modules/phase_field/src/action/DisplacementGradientsAction.C

```diff
@@ -0,0 +1,77 @@
+#include "FEProblem.h"
+
+#include "libmesh/string_to_enum.h"
+
+template<>
+InputParameters validParams<DisplacementGradientsAction>()
+{
+  InputParameters params = validParams<Action>();
+  params.addClassDescription("Set up variables, kernels, and materials for a the displacement gradients and their elastic free energy derivatives for non-split Cahn-Hilliard problems.");
+  params.addRequiredParam<std::vector<VariableName> >("displacements", "Vector of displacement variables");
+  params.addRequiredParam<std::vector<VariableName> >("displacement_gradients", "Vector of displacement gradient variables");
+  params.addParam<Real>("scaling", 1.0, "Specifies a scaling factor to apply to the displacement gradient variables");
+  return params;
+}
+
+DisplacementGradientsAction::DisplacementGradientsAction(const InputParameters & params) :
+    Action(params),
+    _displacements(getParam<std::vector<VariableName> >("displacements")),
+    _displacement_gradients(getParam<std::vector<VariableName> >("displacement_gradients"))
+{
+}
+
+void
+DisplacementGradientsAction::act()
+{
+  unsigned int ngrad = _displacement_gradients.size();
+
+  if (_current_task == "add_variable")
+  {
+    // Loop through the gij variables
+    Real scaling = getParam<Real>("scaling");
+    for (unsigned int i = 0; i < ngrad; ++i)
+    {
+      // Create displacement gradient variables
+      _problem->addVariable(_displacement_gradients[i],
+                            FEType(Utility::string_to_enum<Order>("FIRST"),
+                                   Utility::string_to_enum<FEFamily>("LAGRANGE")),
+                            scaling);
+    }
+  }
+  else if (_current_task == "add_material")
+  {
+    InputParameters params = _factory.getValidParams("StrainGradDispDerivatives");
+    params.set<std::vector<VariableName> >("displacement_gradients") = _displacement_gradients;
+    params.set<std::vector<SubdomainName> >("block") = std::vector<SubdomainName>(1, "0");
+    _problem->addMaterial("StrainGradDispDerivatives", "strain_grad_disp_derivatives", params);
+  }
+  else if (_current_task == "add_kernel")
+  {
+    unsigned int ndisp = _displacements.size();
+    if (ndisp * ndisp != ngrad)
+      mooseError("Number of displacement gradient variables must be the square of the number of displacement variables.");
+
+    // Loop through the displacements
+    unsigned int i = 0;
+    for (unsigned int j = 0; j < ndisp; ++j)
+      for (unsigned int k = 0; k < ndisp; ++k)
+      {
+        InputParameters params = _factory.getValidParams("GradientComponent");
+        params.set<NonlinearVariableName>("variable") = _displacement_gradients[i];
+        params.set<std::vector<VariableName> >("v") = std::vector<VariableName>(1, _displacements[j]);
+        params.set<unsigned int>("component") = k;
+        _problem->addKernel("GradientComponent", _displacement_gradients[i] + "_grad_kernel", params);
+        ++i;
+      }
+  }
+  else
+    mooseError("Internal error.");
```

Comment:
Well, yeah, but this error will only show up if someone registers the action under an additional unsupported task. So this is not a user-facing error but rather a developer-facing error. I can make it more verbose, but it should be rather obvious to a developer what happened...